# COL

**COL** is a novel computational pipeline designed to identify putatively functional circular RNAs (circRNAs) by integrating three key features: higher back-splicing rates, conserved splicing motifs, and unusually high back-splicing levels. Unlike existing methods that require multiple samples, COL can predict functional back-splicing from a single sample, making it highly versatile. Moreover, it outperforms traditional approaches based solely on back-splicing levels by exhibiting a lower false positive rate. This makes COL an efficient tool for the rapid identification of functional circRNAs, streamlining the process for subsequent experimental validation.

For more information, visit [COL GitHub Repository](https://github.com/XuLabSJTU/COL). For details of this work, users can refer to our paper.

## Requirements

To run COL, you need the following software and R packages:

- **R >= 3.0 (64-bit)**
- **BiocManager** 1.30.22
- **Biostrings** 2.68.1
- **liftOver** 1.24.0
- **rtracklayer** 1.62.0
- **BSgenome** 1.68.0
- **GenomeInfoDb** 1.36.4
- **GenomicRanges** 1.54.1
- **IRanges** 2.34.1
- **stringr** 1.5.1
- **devtools** 2.4.5

## Installation

You can install COL and its dependencies as follows:

```r
# Install BiocManager if not already installed
if (!requireNamespace("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}

# Install required Bioconductor packages
BiocManager::install(c("Biostrings","liftOver","rtracklayer", "BSgenome","GenomeInfoDb","GenomicRanges","IRanges"))

# Install required CRAN packages
install.packages(c("stringr","devtools","doParallel","foreach","parallel","data.table"))

# Install and load COL
require("devtools")
devtools::install_github("XuLabSJTU/COL")
```

## Modules

### 1. **outlier**
This module identifies circRNAs that are outliers in the relationship between back-splicing rate and splicing amount.

**Input Files:**
- CIRI_output\: Generated by [CIRI2](https://ciri-cookbook.readthedocs.io/en/latest/CIRI2.html)
- STAR_output\: Splicing junction file generated by [STAR](https://physiology.med.cornell.edu/faculty/skrabanek/lab/angsd/lecture_notes/STARmanual.pdf) (*SJ.out.tab*)

**Parameters:**
- k\: Cutoff for choosing outliers, default is 4 times the mean of Cook’s distance. Adjust to 3 or 2 if the sample has few junctions.

### 2. **cmotif**
This module identifies splicing signals in a species.

### 3. **conserve**
This module combines the above two modules and returns functional circRNA candidates predicted by the COL method.

**Input Files:**
- cirRNA\: Generated by CIRI2.
- STAR_output\: Splicing junction file generated by STAR (*SJ.out.tab*).

**Output Files:**
- ./test/output/FunctionalCirc.txt

**Example Output File:**
| Column | Description |
|--------|-------------|
| 1      | ID of a predicted circRNA in the pattern of \"chr@start@end@strand\". |
| 2      | Circular junction read count of a predicted circRNA. |
| 3      | Sum of circular junction read count of a predicted circRNA's left site. |
| 4      | Sum of circular junction read count of a predicted circRNA's right site. |
| 5      | Sum of linear junction read count of a predicted circRNA's left site. |
| 6      | Sum of linear junction read count of a predicted circRNA's right site. |
| 7      | Total splicing amount of a predicted circRNA. |
| 8      | Back-splicing rate of a predicted circRNA, calculated as 2*back-splicing reads/total splicing amount. |
| 9      | Cook's distance of a predicted circRNA. |
| 10     | Outlier status based on Cook’s distance. (\"1\" for outlier over the regression line, \"-1\" for outlier under the regression line, \"0\" for non-outlier) |
| 11     | Whether or not it has conserved motif between two species. (\"1\" for conserved motif, \"0\" for unconserved motifs) |
| 12     | Rank of back-splicing read count. |
| 13     | Whether or not it is a functional circRNA predicted by COL. |
| 14     | RRA scores calculated by the RobustRankAggreg package. |
| 15     | Rank of RRA scores. |

## Testing

To test the COL pipeline, you can download the DNA sequences of the species of interest.

```r
BiocManager::install(c("BSgenome.Hsapiens.UCSC.hg38","BSgenome.Mmusculus.UCSC.mm10"))
```

> **Note:** If you need to identify other species' functional circRNAs, download their corresponding genome sequences.

Run the COL pipeline as follows:

```r
library("COL")
library("rtracklayer")

# Import chain file
chainObject <- import.chain("./test/hg38ToMm10.over.chain")

# Run the COL pipeline
conserve(cirRNA ="./test/data/SRR13142160CIRI2", 
         STAR_output = "./test/data/SRR13142160SJ.out", 
         genome1 = "Hgenome", 
         genome2 = "Mgenome", 
         chainObject = chainObject, 
         outpath = "./test/output/")
```
